hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializer.java

```diff
@@ -222,24 +222,38 @@ private CompactSerializableRegistration getOrCreateRegistration(Class clazz) {
     }
 
     private CompactSerializableRegistration getOrCreateRegistration(String typeName) {
-        return typeNameToRegistrationMap.computeIfAbsent(typeName, s -> {
-            Class<?> clazz;
-            try {
-                // When the registration does not exist, we treat typeName as className
-                // to check if there is a class with the given name in the classpath.
-                clazz = ClassLoaderUtil.loadClass(classLoader, typeName);
-            } catch (Exception e) {
-                // There is no such class that has typeName as its name.
-                // We should try to read this as GenericRecord. We are
-                // returning this registration here to remember that we
-                // should read instances of this typeName as GenericRecords,
-                // instead of trying to load a class with that name over
-                // and over.
-                return CompactSerializableRegistration.GENERIC_RECORD_REGISTRATION;
-            }
+        CompactSerializableRegistration currentRegistration = typeNameToRegistrationMap.get(typeName);
+        if (currentRegistration != null) {
+            return currentRegistration;
+        }
+        // Execute potentially long-lasting operation outside CHM lock in computeIfAbsent.
+        // Some special classloaders (eg. JetClassLoader) may try to access external resources
+        // and require other threads.
+        // We might try to load the same class multiple times in parallel but this is not a problem.
+        CompactSerializableRegistration newRegistration = getOrCreateRegistration0(typeName);
+
+        // Registration might have been created by a concurrent thread.
+        // If so, use that one instead.
+        return typeNameToRegistrationMap.computeIfAbsent(typeName, k -> newRegistration);
+    }
 
-            return getOrCreateRegistration(clazz);
-        });
+    private CompactSerializableRegistration getOrCreateRegistration0(String typeName) {
+        Class<?> clazz;
+        try {
+            // When the registration does not exist, we treat typeName as className
+            // to check if there is a class with the given name in the classpath.
+            clazz = ClassLoaderUtil.loadClass(classLoader, typeName);
+        } catch (Exception e) {
+            // There is no such class that has typeName as its name.
+            // We should try to read this as GenericRecord. We are
+            // returning this registration here to remember that we
+            // should read instances of this typeName as GenericRecords,
+            // instead of trying to load a class with that name over
+            // and over.
+            return CompactSerializableRegistration.GENERIC_RECORD_REGISTRATION;
```

Comment:
Is it actually safe to cache this indefinitely? 

E.g. you have generic record and the class is not on classpath, and then you re-deploy new version of the job, and the class might be there. Would that cause issues?

Reply:
yes, if the Compact class is uploaded in the job then we would have memory leak because of keeping the reference to Jet classloader in `CompactStreamSerializer.classToRegistrationMap`. I forgot about it, but similar leak was reported on [community slack](https://hazelcastcommunity.slack.com/archives/C013LM6JK7S/p1688542710937219). But that is a separate issue, not affected by this PR.

Reply:
maybe someone with knowledge of the design could answer the question if it is allowed for compact classes to be uploaded in Jet job. Current implementation would cause problems with that, but this might be just unsupported config without enough validation on member side.

Reply:
> loading Compact classes from Jet job jars is a very bad idea
I don't think it is - we say compact is the serialization you should use and if you need to create a POJO in the pipeline, it needs to be somehow serialized and using compact is valid

Can we create a separate issue for the possible leak?

Reply:
https://hazelcast.atlassian.net/browse/HZ-3121

Reply:
using attached compact classes is not trivial to implement